core/drivers/crypto/caam/utils/utils_dmaobj.c

```diff
@@ -0,0 +1,1405 @@
+		goto out;
+	}
+
+	if ((input && !input->priv) || (output && !output->priv)) {
+		ret = TEE_ERROR_BAD_PARAMETERS;
+		goto out;
+	}
+
+	DMAOBJ_TRACE("input=%p - output=%p - min=%zu", input, output, min_size);
+
+	ret = try_allocate_dmabuf_max_size(input, min_size, &alloc_input);
+	if (ret)
+		goto out;
+
+	ret = try_allocate_dmabuf_max_size(output, min_size, &alloc_output);
+	if (ret)
+		goto out;
+
+out:
+	DMAOBJ_TRACE("Allocation (input %zu, output %zu) returns 0x%" PRIx32,
+		     input ? alloc_input : 0, output ? alloc_output : 0,
+		     ret);
+
+	return ret;
+}
+
+TEE_Result caam_dmaobj_sgtbuf_inout_build(struct caamdmaobj *input,
+					  struct caamdmaobj *output,
+					  size_t *length, size_t off,
+					  size_t align)
+{
+	TEE_Result ret = TEE_ERROR_GENERIC;
+	size_t len = 0;
+
+	DMAOBJ_TRACE("input=%p/output=%p %zu bytes (offset=%zu, align=%zu)",
+		     input, output, *length, off, align);
+
+	if (!input || !output || !length || !input->priv || !output->priv ||
+	    !*length) {
+		ret = TEE_ERROR_BAD_PARAMETERS;
+		goto out;
+	}
+
+	/*
+	 * First build the input SGT/Buffer
+	 */
+	ret = caam_dmaobj_sgtbuf_build(input, length, off, align);
+	if (ret)
+		goto out;
+
+	/*
+	 * Next build the output SGT/Buffer.
+	 * If returned length is not same as input, redo the input
+	 * SGT/Buffer with the same length as the output.
+	 */
+	len = *length;
+	ret = caam_dmaobj_sgtbuf_build(output, &len, off, *length);
+	if (ret)
+		goto out;
+
+	if (len != *length) {
+		DMAOBJ_TRACE("Retry In %zu bytes vs Out %zu bytes", *length,
+			     len);
+
+		/* Redo the input with the output length */
+		*length = len;
+		ret = caam_dmaobj_sgtbuf_build(input, length, off, len);
+		if (!ret && *length != len) {
+			DMAOBJ_TRACE("Error In %zu bytes vs Out %zu bytes",
+				     *length, len);
+			ret = TEE_ERROR_OUT_OF_MEMORY;
+		}
+	}
+
+out:
+	DMAOBJ_TRACE("Input/Output SGTBUF returns 0x%" PRIx32, ret);
+
+	return ret;
+}
+
+TEE_Result caam_dmaobj_sgtbuf_build(struct caamdmaobj *obj, size_t *length,
+				    size_t off, size_t align)
+{
+	TEE_Result ret = TEE_ERROR_GENERIC;
+	enum caam_status retstatus = CAAM_FAILURE;
+	struct priv_dmaobj *priv = NULL;
+	struct dmaentry *entry = NULL;
+	struct dmaentry *start_entry = NULL;
+	size_t max_length = 0;
+	size_t acc_length = 0;
+	size_t offset = off;
+	unsigned int idx = 0;
+	unsigned int nb_sgt = 0;
+
+	DMAOBJ_TRACE("obj=%p of %zu bytes (offset=%zu) - align %zu", obj,
+		     *length, off, align);
+
+	if (!obj || !obj->priv || !length || !*length) {
+		ret = TEE_ERROR_BAD_PARAMETERS;
+		goto out;
+	}
+
+	priv = obj->priv;
+
+	max_length = *length;
+	if (priv->dmabuf.allocated && max_length > priv->dmabuf.allocated &&
+	    priv->dmabuf.allocated > align)
+		max_length = ROUNDDOWN(priv->dmabuf.allocated, align);
+
+	DMAOBJ_TRACE("Prepare SGT/Buffer to do %zu of %zu", max_length,
+		     *length);
+
+	/* Find the first DMA buffer to start with */
+	TAILQ_FOREACH(entry, &priv->list, link)
+	{
```

Comment:
Please move this to the end of previous line.

Reply:
I missed this one, I will fix it